torchrl/collectors/distributed/distributed_collector.py

```diff
@@ -0,0 +1,232 @@
+                 ):
+
+        if not _has_ray:
+            raise RuntimeError(
+                f"ray library not found, unable to create a DistributedCollector. "
+            )
+
+        if communication not in ("sync", "async"):
+            raise ValueError(f"Communication parameter in CollectorSet has to be sync or async.")
+
+        # Monkey patching as_remote and print_remote_collector_info to collector class # TODO: is that ok ?
+        collector_class.as_remote = as_remote
+        collector_class.print_remote_collector_info = print_remote_collector_info
+
+        self.collected_frames = 0
+        self.total_frames = total_frames
+        self.collector_class = collector_class
+        self.collector_params = collector_params
+        self.num_collectors = num_collectors
+        self.remote_config = remote_config
+        self.communication = communication
+
+        # Create a local instance of the collector class.
+        # Will be used to track latest policy weights.
+        self._local_collector = self._make_collector(
+            self.collector_class, collector_params)
+
+        # Create remote instances of the collector class
+        self._remote_collectors = []
+        if self.num_collectors > 0:
+            self.add_collectors(self.num_collectors, collector_params)
+
+        # Print info of all remote workers
+        pending_samples = [e.print_remote_collector_info.remote() for e in self.remote_collectors()]
+        ray.wait(pending_samples)
+
+    @staticmethod
+    def _make_collector(cls, collector_params):
+        """Create a single collector instance."""
+        w = cls(**collector_params)
+        return w
+
+    def add_collectors(self, num_collectors, collector_params):
+        """Create and add a number of remote collectors to the set."""
+        cls = self.collector_class.as_remote(**self.remote_config).remote
+        self._remote_collectors.extend(
+            [self._make_collector(cls, collector_params) for _ in range(num_collectors)])
+
+    def local_collector(self):
+        """Return local collector"""
+        return self._local_collector
+
+    def remote_collectors(self):
+        """Returns list of remote collectors"""
+        return self._remote_collectors
+
+    def stop(self):
+        """Stop all remote collectors"""
+        for w in self.remote_collectors():
+            w.__ray_terminate__.remote()
+
+    def __iter__(self) -> Iterator[TensorDictBase]:
+        if self.communication == "sync":
+            return self.sync_iterator()
+        else:
+            return self.async_iterator()
+
+    def sync_iterator(self) -> Iterator[TensorDictBase]:
+
+        while self.collected_frames < self.total_frames:
+
+            # Broadcast agent weights
+            self.local_collector().update_policy_weights_()
+            state_dict = {"policy_state_dict": self._local_collector.policy.state_dict()}
+            state_dict = ray.put(state_dict)
```

Comment:
Q: in regular data collectors we leave this to the user or ask explicitly if the update should be done at each iter.
Do we want to do the same here? ie isolating the weight updating in a dedicated method that can or won't be called at each iteration.